[Shawn Kok] iP

Kok-je opened the pull request:
# Sakura

> "For all x there exist y in life such that y is more brutal than x." - [Nguyen Vu Khang 💀](https://github.com/nguyenvukhang) 

Ever wanted a *cute waifu* that understands you deeply and helps you with your *daily tasks*? Well, the future is now old man!

Introducing Sakura, your best virtual companion!

## Capabilities

Sakura is perhaps the greatest innovation the human race bore witness to since the [Virtual Hologram Girlfriend](https://cydskd.jpn.org/about).

- Keeping track of your tasks
- Keeping track of deadlines
- Storing and deleting as many tasks as you will ever need in a hundred lifetimes
- Greeting you
- Scolding you

## Setup

The installation and setup is easy. All you need to do is this:
1. Download the latest jar file from [here](https://github.com/Kok-je/ip/releases)
2. Run the jar file with the following command
    
bash
    java -jar /path/to/jar/file
    
3. You're all ready to use Sakura! 🔥 

## Roadmap

- [X] Basic Task Manager Functions
- [x] Generative Chatting Functionality
- [ ] ~~Conquer the World~~ World Peace

## Contributing

Collaborations are currently closed. Please come back again next season!
_Excerpt from Sakura source code_
```
java
public class Sakura{
    /**
     * Code entry point.
     *
     * @param args Entry point argument.
     */
    public static void main(String[] args) throws IndexOutOfBoundsException {
        new Sakura("./data/Sakura.txt").run();
    }
}
```

Commit:
Add Gradle support

Kok-je committed:
Level 1. Greet, Echo, Exit

Kok-je committed:
Level 2. Add, List

Kok-je committed:
Leve 3. Mark as Done

Kok-je committed:
Level 4. ToDos, Events, Deadlines

Kok-je committed:
Level 5. Handle Errors

Kok-je committed:
Level 6. Delete

Kok-je committed:
A-TextUiTesting

Kok-je committed:
Level 7. Save

Kok-je committed:
Level 8. Dates and Times

Kok-je committed:
Renamed TaskManager to TaskList

Kok-je committed:
Added greet, exit, addDescription, deleteDescription, showAllTask to UI

Kok-je committed:
Added loadData, readEntry, createDatabase, saveData to Storage

Kok-je committed:
Added parseCommand and exit to Parser, refactored code

Kok-je committed:
removed extra imports

Kok-je committed:
A-Packages

Kok-je committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Kok-je committed:
Misc updates

Kok-je committed:
gitignore

Kok-je committed:
Merge branch 'branch-Level-7' into HEAD

Kok-je committed:
Merge branch 'branch-Level-8' into HEAD

Kok-je committed:
Temp

Kok-je committed:
update gitignore

Kok-je committed:
fix merge conflict

Kok-je committed:
Fix merge conflict

Kok-je committed:
A-JUnit

Kok-je committed:
A-Jar

Kok-je committed:
A-JavaDoc

Kok-je committed:
Merge branch 'branch-A-JavaDoc'

Kok-je committed:
A-CodingStandard

Kok-je committed:
Merge tag 'A-CodingStandard'

Kok-je committed:
Level-9. Find

Kok-je committed:
Merge branch 'branch-Level-9'

Kok-je committed:
Fixed Ui errors

jetlfj commented on src/main/java/Sakura/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    /**
+     * Convert the deadline into the specific string for data saving.
```
Remember to use the correct form.
```suggestion
     * Converts the deadline into the specific string for data saving.
```

jetlfj commented on src/main/java/Sakura/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * Return the string representation of the deadline.
+     *
+     * @return string representation of the deadline
```
Remember the punctuation behind each parameter description.
```suggestion
     * @return String representation of the deadline.
```

jetlfj commented on src/main/java/Sakura/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+package Sakura;
```
Shouldn't package names be lowercase to avoid confusion with class names?

jetlfj commented on src/main/java/Sakura/Storage.java:
```diff
@@ -0,0 +1,108 @@
+package Sakura;
+
+import java.io.*;
```
Shouldn't all import classes be listed explicitly?

jetlfj commented on src/main/java/Sakura/TaskList.java:
```diff
@@ -0,0 +1,154 @@
+package Sakura;
+import java.time.format.DateTimeParseException;
```
Shouldn't there be an empty line between these two lines?
```suggestion
package Sakura;

import java.time.format.DateTimeParseException;
```

jetlfj commented on src/main/java/Sakura/Parser.java:
```diff
@@ -0,0 +1,51 @@
+     */
+    public static void parseCommand(String input, TaskList taskList) {
+        Sakura.ui.printDiv();
+        if (input.equals("bye")) {
+            Sakura.inProgress = false;
+            exit();
+        } else if (input.equals("list")) {
+            Ui.showAllTask(taskList.tasks);
+        } else if (input.toLowerCase().startsWith("mark")) {
+            taskList.markTask(input);
+        } else if (input.toLowerCase().startsWith("unmark")) {
+            taskList.unmarkTask(input);
+        } else if (input.toLowerCase().startsWith("todo") || input.toLowerCase().startsWith("deadline") || input.toLowerCase().startsWith("event")) {
+            taskList.addTask(input);
+        } else if (input.toLowerCase().startsWith("delete")) {
+            taskList.deleteTask(input);
+        } else if (input.toLowerCase().startsWith("find")) {
+            taskList.searchTask(input);
+        } else {
+            SakuraException.genericTask();
+        }
```
Consider putting the toLowerCase() at the start by retrieving the first word in the input and using a switch statement. This makes it:
1. Easier to read.
2. Accept Bye and List commands which are not lowercase.

jetlfj commented on src/main/java/Sakura/Ui.java:
``````````diff
@@ -0,0 +1,105 @@
+}
+
+
+//               "\u001B[31m███████  █████  ██   ██ ██    ██ ██████   █████    \u001B[0m\n"
+//                       + "\u001B[31m██      ██   ██ ██  ██  ██    ██ ██   ██ ██   ██   \u001B[0m\n"
+//                       + "\u001B[31m███████ ███████ █████   ██    ██ ██████  ███████   \u001B[0m\n"
+//                       + "\u001B[31m     ██ ██   ██ ██  ██  ██    ██ ██   ██ ██   ██   \u001B[0m\n"
+//                       + "\u001B[31m███████ ██   ██ ██   ██  ██████  ██   ██ ██   ██  ██\u001B[0m\n";
+
+//            "\u001B[36m            ..''''       .'.       |    ..'' |         | |`````````,       .'.       \u001B[0m\n"
+//                    + "\u001B[36m         .''           .''```.     |..''     |         | |'''|'''''      .''```.     \u001B[0m\n"
+//                    + "\u001B[34m      ..'            .'       `.   |``..     |         | |    `.       .'       `.   \u001B[0m\n"
+//                    + "\u001B[35m....''             .'           `. |    ``.. `._______.' |      `.   .'           `. \u001B[0m\n";
``````````
Could be removed if no longer needed.

jetlfj reviewed (commented):
Overall, the code was easy to read. Naming of variables and methods seems to be mostly on point. I noted a few minor coding standard violations, mainly the formatting of imports and Javadoc comments throughout the code.

Kok-je committed:
A-CheckStyle

yeojunjie commented on src/main/java/Sakura/Event.java:
```diff
@@ -0,0 +1,46 @@
+ */
+public class Event extends Task {
+
+    protected LocalDateTime at;
```
Hi Shawn, it was not immediately obvious what `at` meant. I would like to suggest using a more descriptive name, such as `startTime`.
```suggestion
    protected LocalDateTime startTime;
```

yeojunjie commented on src/main/java/Sakura/Ui.java:
```diff
@@ -0,0 +1,105 @@
+     * @param tasks current list of tasks.
+     * @param deletedTask task to be deleted.
+     */
+    public static void deleteDescription(List<Task> tasks, Task deletedTask) {
```
Hi Shawn, separating the operands to `+` in `System.out.println()` and aligning them neatly helped me understand what was meant to be printed in one glance.

However, I think this method (and similar methods) could be better named as `displayDeleteDescription`.

Renaming `tasks` to `remainingTasks` helps convey that this method is only called after the task has been deleted (though it can admittedly be inferred from the context).

```suggestion
    public static void displayDeleteDescription(List<Task> remainingTasks, Task deletedTask) {
```

yeojunjie commented on src/main/java/Sakura/Ui.java:
```diff
@@ -0,0 +1,105 @@
+        System.out.println(list);
+    }
+
+    public static void searchTaskDescription(List<Task> searchTask) {
```
Similarly, I suggest renaming the method and argument.

```suggestion
    public static void displaySearchResults(List<Task> searchResults) {
```

yeojunjie commented on src/main/java/Sakura/TaskList.java:
```diff
@@ -0,0 +1,154 @@
+                if (input.split(" ").length == 1) {
+                    SakuraException.invalidTodo();
+                } else {
+                    Todo todo = new Todo(input.substring("todo ".length()));
```
I like how `"todo ".length()` was not replaced with the magic number `5`!

This helps me understand that we are getting the name of the task by omitting the first five characters.

I should adopt this in my code as well.

yeojunjie commented on src/main/java/Sakura/Todo.java:
```diff
@@ -0,0 +1,36 @@
+     * @return the string representation in the specific format for database
+     */
+    @Override
+    public String stringifyTask() {
```
```suggestion
    public String encodeTask() {
```

Hi Shawn, I think this renaming will help to communicate how `stringifyTask` is different from `toString`.

I liked the use of `String.format()`. It helped me immediately understand the encoding scheme. In comparison, the alternative would be `return ("T|" + this.isDone ? "1" : "0" + "|" + this.description)`, which is nowhere as elegant.

yeojunjie reviewed (commented):
Overall, the code was very easy to read because the indentation and whitespace rules are observed very carefully. There were some instances where methods and fields could be named even more clearly!

Kok-je committed:
Level 10. GUI

Kok-je committed:
Merge branch 'branch-Level-10'

Kok-je committed:
Update GUI interface

Kok-je committed:
Update GUI interface

Kok-je committed:
Updates in UI

Kok-je committed:
A-Assertions

Kok-je committed:
A-CodeQuality: Improve Code Quality

Kok-je committed:
C-Sort

Kok-je committed:
Merge pull request #4 from Kok-je/branc-C-Sort

C-Sort

Kok-je committed:
Merge pull request #2 from Kok-je/branch-A-Assertions

A-Assertions

Kok-je committed:
Merge pull request #3 from Kok-je/branch-A-CodeQuality

A-CodeQuality: Improve Code Quality

Kok-je committed:
Removed unecessary code.

Kok-je committed:
Refactor Task.java TaskList.java Todo.java Deadline.java Event.java to task inner package

There are several way to practice better code quality. Using packages and folders is one of them.

Refactoring will make searching or locating classes easier, implements data encapsulation, and provides controlled access.

Let's,
* refactor related java class files under a unified package called task
* edit the access modifiers of affected java files
Doing so will make our code more readable.

Kok-je committed:
A-BetterGui, A-Personality, A-MoreErrorHandling, A-MoreTesting

Kok-je committed:
User Guide

Kok-je committed:
Update Readme

Kok-je committed:
update dependencies

Kok-je committed:
update docs

Kok-je committed:
Update Readme

Kok-je committed:
Update README.md

Kok-je committed:
Update README.md
